oauth2_http/java/com/google/auth/oauth2/GdchCredentials.java

```diff
@@ -0,0 +1,463 @@
+    try {
+      KeyFactory keyFactory = SecurityUtils.getRsaKeyFactory();
+      return keyFactory.generatePrivate(keySpec);
+    } catch (NoSuchAlgorithmException | InvalidKeySpecException exception) {
+      unexpectedException = exception;
+    }
+    throw new IOException("Unexpected exception reading PKCS#8 data", unexpectedException);
+  }
+
+  /**
+   * Create a copy of GDCH credentials with the specified audience.
+   *
+   * @param apiAudience The intended audience for GDCH credentials.
+   */
+  public GdchCredentials createWithGdchAudience(URI apiAudience) throws IOException {
+
+    if (apiAudience == null) {
+      throw new IOException("Audience are not configured for GDCH service account credentials.");
+    }
+    return this.toBuilder().setGdchAudience(apiAudience).build();
+  }
+
+  /** Refresh the OAuth2 access token by getting a new access token using a JSON Web Token (JWT). */
+  @Override
+  public AccessToken refreshAccessToken() throws IOException {
+
+    if (this.apiAudience == null) {
+      throw new IOException(
+          "Audience are not configured for GDCH service account. Specify the "
+              + "audience by calling createWithGDCHAudience.");
+    }
+
+    JsonFactory jsonFactory = OAuth2Utils.JSON_FACTORY;
+    long currentTime = clock.currentTimeMillis();
+    String assertion = createAssertion(jsonFactory, currentTime, getApiAudience());
+
+    GenericData tokenRequest = new GenericData();
+    tokenRequest.set("grant_type", GRANT_TYPE);
+    tokenRequest.set("assertion", assertion);
+    UrlEncodedContent content = new UrlEncodedContent(tokenRequest);
+
+    HttpRequestFactory requestFactory = transportFactory.create().createRequestFactory();
+    HttpRequest request = requestFactory.buildPostRequest(new GenericUrl(tokenServerUri), content);
+
+    request.setParser(new JsonObjectParser(jsonFactory));
+
+    HttpResponse response;
+    String errorTemplate = "Error getting access token for GDCH service account: %s, iss: %s";
+
+    try {
+      response = request.execute();
+    } catch (HttpResponseException re) {
+      String message = String.format(errorTemplate, re.getMessage(), getServiceIdentityName());
+      throw GoogleAuthException.createWithTokenEndpointResponseException(re, message);
+    } catch (IOException e) {
+      throw GoogleAuthException.createWithTokenEndpointIOException(
+          e, String.format(errorTemplate, e.getMessage(), getServiceIdentityName()));
+    }
+
+    GenericData responseData = response.parseAs(GenericData.class);
+    String accessToken =
+        OAuth2Utils.validateString(responseData, "access_token", PARSE_ERROR_PREFIX);
+    int expiresInSeconds =
+        OAuth2Utils.validateInt32(responseData, "expires_in", PARSE_ERROR_PREFIX);
+    long expiresAtMilliseconds = clock.currentTimeMillis() + expiresInSeconds * 1000L;
+    return new AccessToken(accessToken, new Date(expiresAtMilliseconds));
+  }
+
+  /** Create a self-signed JWT. */
```

Comment:
Also could be an explanation of the GDCH flow. The usage of self-signed JWT is unique and used for token exchange, not for API call.

Reply:
Updated the documentation.